content/blog/2022-05-17-object-oriented-features/index.md

```diff
@@ -0,0 +1,210 @@
+
+We opted for approach 2 for simplicity, although approach 2 is definetly the better approach. We discuss this more in future work section.
+
+# Future Work
+
+## Typechecking
+
+This project skipped over the type checking phase so that more time could be spent focusing on the actual object oriented features. Obviously, this allows Xi programs that are not well typed and would likley crash or misbehave if these types of programs are used. Additionally, type checking would allow programs that involve more than one class. At the moment, it is impossible to figure out what class a variable `o` belongs to in situations such as `o.f()` and `o.x`. Currently, they are hardcoded to assume that all objects are of type `A`. Type checking would allow the compiler to identify what type a variable has and therefore make decisions based on class layout for any class type.
+
+## Fields
+
+All fields are private to a class. This means that if a class `A` has field `x` and class `B` entends `A`, then class `B` cannot refer to `x`. However, `B` can call methods of `A` and those methods may refer to `x`. Currently, subclasses do not allocate fields that belong to their parent classes. Therefore when a method of a super class is called, it has no way to get those fields. In the current implementation it will try and get the variable from the expected index which will really be an
+undefined memory access.
+
+In order to remedy this problem, we must allocate fields belonging to super classes at runtime since fields are not part of class declarations and are only part of class definitions by the Xi++ specification. I spoke with Professer Myers about this. C++ solves this problem by requiring public and private fields to be declared in the header files (the class declaration) so this information is known at compile time. Java on the otherhand does this at runtime by having a bytecode instruction that refers to the variables by name. In Xi, the runtime would probably need to do something similiar to the dispatch vector indexing scheme where all variables of a subtype are placed at a certain index in the object layout at runtime.
+
+## Static Dispatch Vectors
+
+As discussed above, much of the work that went into this project was with respect to the representation of dispatch vectors in the IR. I met with Professer Myers and he agreed that his intention is for these dispatch vectors to be static data but there is no way to represent them using the current version of `IRData` since `IRData` requires the physical address of functions but these are not determined until runtime. The solution is to create a different `IRData` class which can represent static data that is of the form `IRNode` instead of physical addresses. This would allow us to say the data is an array of `IRName`s which would be converted to the address of the correct function at runtime.
```

Comment:
Care to comment on how this approach compares to what you actually did, in terms of effectiveness and effort?